### ISSUE TITLE
INTERNAL ASSERT FAILED when loading saved Python model in C++

### ISSUE THREAD

**Issue by ankane** [pytorch/pytorch#42651]
## 🐛 Bug

Hi, I'm trying to load a saved ResNet model from torchvision in C++ (after converting it to the new zip file format), but an internal assertion is failing when loading the model.

## To Reproduce

Download ResNet18 from torchvision

```sh
wget https://download.pytorch.org/models/resnet18-5c106cde.pth
```

Convert the file to the new zip file format with Python

```python
import torch

torch.save(torch.load("resnet18-5c106cde.pth"), "resnet18.pth")
```
 
Build the following C++ application

```cpp
#include <torch/torch.h>
#include <iostream>
#include <iterator>

int main() {
  std::ifstream file("../resnet18.pth", std::ios::binary);
  std::vector<char> data((std::istreambuf_iterator<char>(file)), std::istreambuf_iterator<char>());
  torch::IValue ivalue = torch::pickle_load(data);
}
```

Run it and it'll fail with

```txt
libc++abi.dylib: terminating with uncaught exception of type c10::Error: type_resolver_ INTERNAL ASSERT FAILED at "../torch/csrc/jit/serialization/unpickler.cpp":588, please report a bug to PyTorch. 
Exception raised from readGlobal at ../torch/csrc/jit/serialization/unpickler.cpp:588 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >) + 169 (0x100cbb7f9 in libc10.dylib)
frame #1: torch::jit::Unpickler::readGlobal(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) + 8427 (0x1048cfb6b in libtorch_cpu.dylib)
frame #2: torch::jit::Unpickler::readInstruction() + 3351 (0x1048c85c7 in libtorch_cpu.dylib)
frame #3: torch::jit::Unpickler::run() + 472 (0x1048c7528 in libtorch_cpu.dylib)
frame #4: torch::jit::Unpickler::parse_ivalue() + 34 (0x1048c7082 in libtorch_cpu.dylib)
frame #5: torch::jit::readArchiveAndTensors(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, c10::optional<std::__1::function<c10::StrongTypePtr (c10::QualifiedName const&)> >, c10::optional<std::__1::function<c10::intrusive_ptr<c10::ivalue::Object, c10::detail::intrusive_target_default_null_type<c10::ivalue::Object> > (c10::StrongTypePtr, c10::IValue)> >, c10::optional<c10::Device>, caffe2::serialize::PyTorchStreamReader&) + 838 (0x104867d76 in libtorch_cpu.dylib)
frame #6: torch::jit::pickle_load(std::__1::vector<char, std::__1::allocator<char> > const&) + 347 (0x10489a6bb in libtorch_cpu.dylib)
frame #7: torch::pickle_load(std::__1::vector<char, std::__1::allocator<char> > const&) + 14 (0x104aba90e in libtorch_cpu.dylib)
frame #8: main + 140 (0x100c7f93c in example-app)
frame #9: start + 1 (0x7fff672ffcc9 in libdyld.dylib)
frame #10: 0x0 + 1 (0x1 in ???)
```

## Expected behavior

No crash

## Environment

 - PyTorch Version (e.g., 1.0): 1.6.0 (for both PyTorch and LibTorch)
 - OS (e.g., Linux): Mac OSX 10.15.6
 - How you installed PyTorch (`conda`, `pip`, source): pip for PyTorch, downloaded LibTorch from pytorch.org
 - Build command you used (if compiling from source): 
 - Python version: 3.8
 - CUDA/cuDNN version: n/a
 - GPU models and configuration: n/a
 - Any other relevant information: n/a

cc @suo @gmagogsfm @yf225 @glaringlee

**Comment by glaringlee** [comment]
The latest master branch has this issue as well.

I saw jit does test this feature. 

It saved an ivalue.pt file here:
https://github.com/pytorch/pytorch/blob/master/test/cpp/jit/tests_setup.py#L43

and test loading it here:
https://github.com/pytorch/pytorch/blob/e7a09b4d17010e60bc95c25f8165ef479d1b9612/test/cpp/jit/test_misc.cpp#L509

JIT team might know this better, change label to jit.

**Comment by wanchaol** [comment]
You could use `torch::load` and it should work, we will address the `pickle_load` error that introduced in this issue.

**Comment by ankane** [comment]
Thanks @wanchaol! I'm having trouble getting the program to compile with `torch::load` and `torch::jit::load`. If it's not too much trouble, what should the example above look like?

fwiw, I found the `pickle_load` approach here: https://github.com/pytorch/pytorch/issues/20356#issuecomment-567663701. I'm not sure if there's official documentation for how to share tensors and models between Python and C++.

**Comment by dfalbel** [comment]
I have a similar issue. I opened this post in the forum before finding this: https://discuss.pytorch.org/t/reading-tensors-in-c-saved-with-torch-save/93130

**Comment by glaringlee** [comment]
> I have a similar issue. I opened this post in the forum before finding this: https://discuss.pytorch.org/t/reading-tensors-in-c-saved-with-torch-save/93130

@dfalbel Your post is not related to this issue. I think you forget to add std::noskipws, I have the full answer in your post, please check.

**Comment by dfalbel** [comment]
@glaringlee yes, you are right! thanks for your answer. And I can confirm that whenever the saved dictionary includes a `torch.nn.parameter.Parameter` I get the same error as @ankane . simply converting parameters into pure tensors fixes.

**Comment by ankane** [comment]
Cleaning up stale issues
